src/symbols_linux.cpp

```diff
@@ -690,91 +698,67 @@ void Symbols::parseKernelSymbols(CodeCache* cc) {
-            image_base = map_start;
-            last_inode = inode;
+        if (_parsed_inodes.find(inode) != _parsed_inodes.end()) {
+            continue;  // shared object is already parsed
         }
-
-        if (!map.isExecutable() || !_parsed_libraries.insert(map_start).second) {
-            // Not an executable segment or it has been already parsed
-            continue;
+        if (inode == 0 && strcmp(map.file(), "[vdso]") != 0) {
+            continue;  // all shared libraries have inode, except vDSO
         }
 
+        const char* map_start = map.addr();
         const char* map_end = map.end();
-        if (inode != 0 && !_parsed_inodes.insert(inode).second) {
-            // Do not parse the same executable twice
-            if (inode == cc_inode) {
-                cc->updateBounds(map_start, map_end);
-            }
-            continue;
-        }
-
-        int count = array->count();
-        if (count >= MAX_NATIVE_LIBS) {
-            break;
+        if (inode != last_inode && map.offs() == 0) {
+            image_base = map_start;
+            last_inode = inode;
         }
 
-        cc = new CodeCache(map.file(), count, false, map_start, map_end);
-        cc_inode = inode;
-
-        if (strchr(map.file(), ':') != NULL) {
-            // Do not try to parse pseudofiles like anon_inode:name, /memfd:name
-        } else if (inode != 0) {
-            if (inode == last_inode) {
-                // If last_inode is set, image_base is known to be valid and readable
-                ElfParser::parseFile(cc, image_base, map.file(), true);
-                // Parse program headers after the file to ensure debug symbols are parsed first
-                ElfParser::parseProgramHeaders(cc, image_base, map_end, OS::isMusl());
+        if (map.isExecutable()) {
+            SharedLibrary& lib = libs[inode];
+            if (lib.file == nullptr) {
+                lib.file = strdup(map.file());
+                lib.map_start = map_start;
+                lib.map_end = map_end;
+                lib.image_base = inode == last_inode ? image_base : NULL;
+                max_count--;
             } else {
-                // Unlikely case when image_base has not been found.
-                // Be careful: executable file is not always ELF, e.g. classes.jsa
-                ElfParser::parseFile(cc, map_start, map.file(), true);
+                // The same library may have multiple executable segments mapped
+                lib.map_end = map_end;
             }
-        } else if (strcmp(map.file(), "[vdso]") == 0) {
-            ElfParser::parseProgramHeaders(cc, map_start, map_end, true);
         }
-
-        cc->sort();
-        applyPatch(cc);
-        array->add(cc);
     }
 
     free(str);
     fclose(f);
-
-    return 1;
 }
 
 void Symbols::parseLibraries(CodeCacheArray* array, bool kernel_symbols) {
     MutexLocker ml(_parse_lock);
 
+    if (array->count() >= MAX_NATIVE_LIBS) {
+        return;
```

Comment:
Shouldn't we log something here instead of skipping quietly?

Reply:
Good idea. Added one-time warning.